Fix Tree Borrows & Stacked Borrows Violations

DropDemBits opened the pull request:
Fixes #6

This resolves all TB violations, and with [`6a64bfd`](https://github.com/ColinFinck/nt-list/commit/6a64bfd8fa75e04be0f5d50133f2690f39e3abbc) also resolves all SB violations. I separated out the commit fixing the SB violations since it's still sound under TB without it, and in the event that a TB or TB-like model becomes the default for miri, that commit can easily be reverted.

This also adds miri to the CI workflow to hopefully catch any future UB.

DropDemBits committed:
Add tests modifying cleared lists

Should still not cause undefined behaviour even after clearing the list.

DropDemBits committed:
Derive mutable pointers from mutable references

Deriving a mutable pointer from an immutable reference results in a `Frozen`
mutable pointer, which is UB to write through. Since all of the contexts
that a mutable pointer is needed also have a mutable reference available,
we can derive the pointer from the mutable reference instead.

This fixes TB violations.

DropDemBits committed:
Compare against `end_marker{,_mut}` in list iterators

Makes it more clear that we're comparing against the end marker without needing
to think through what the cast is doing.

DropDemBits committed:
Fix out-of-range accesses from getting the containing record

The issue with the original `containing_record{,_mut}` methods under SB is that
the pointer derived from the list entry reference is only valid for the range
of list entry, and not the containing record itself. By deferring the raw
pointer dereference to once we have a pointer to the containing record, the
resultant reference is valid for the whole record.

This fixes all violations under SB.

DropDemBits committed:
Use miri in CI

ColinFinck commented on nt-list/src/list/base.rs:
```diff
@@ -457,17 +459,22 @@ where
-        unsafe { &*self.element_ptr() }
+    pub(crate) unsafe fn containing_record<'a>(ptr: *const Self) -> &'a E {
+        unsafe { &*Self::element_ptr(ptr) }
     }
 
-    pub(crate) fn containing_record_mut(&mut self) -> &mut E {
-        unsafe { &mut *(self.element_ptr() as *mut E) }
+    pub(crate) unsafe fn containing_record_mut<'a>(ptr: *mut Self) -> &'a mut E {
+        unsafe { &mut *Self::element_ptr_mut(ptr) }
     }
```
`element_ptr()` has originally been written as a common function for `containing_record()` and `containing_record_mut()`.
Now that `containing_record()` uses `element_ptr()`, and `containing_record_mut()` uses `element_ptr_mut()`, and both are the only customers of the "element_ptr" functions, it makes sense to move it all to `containing_record()`/`containing_record_mut()` and get rid of `element_ptr()`/`element_ptr_mut()`.

DropDemBits replied:
Inlining `element_ptr_{mut}()` also makes the implementation nicely parallel with the `Nt{Single}ListHead::entry` function, so sounds good to me either way.

ColinFinck commented on nt-list/src/list/base.rs:
```diff
@@ -73,47 +73,49 @@ where
-    pub(crate) fn end_marker(self: Pin<&Self>) -> *mut NtListEntry<E, L> {
+    /// Returns a const pointer to the "end marker element" (which is the address of our own `NtListHead`, but interpreted as a `NtListEntry` element address).
+    pub(crate) fn end_marker(self: Pin<&Self>) -> *const NtListEntry<E, L> {
         (self.get_ref() as *const _ as *mut Self).cast()
```
```suggestion
        (self.get_ref() as *const Self).cast()
```

Now that `end_marker()` outputs a `*const NtListEntry`, we don't need another intermediate conversion step, even less a mutable one.

ColinFinck reviewed (changes requested):
Thank you very much @DropDemBits for looking into this, and coming up with such a detailed PR and comprehensible commit messages!
I know for a long time that I needed to look into this, but didn't find the time so far.

I have reviewed your PR and found just two minor things.
Please let me know what you think about the outstanding suggestions.

DropDemBits committed:
Remove mutable cast in `end_marker`

This was a left-over from when `end_marker` was used to also get a mutable
pointer to the end marker. The extra mutable cast isn't needed as it gets
coerced back into an immutable pointer, so we can remove it.

DropDemBits committed:
Inline usages of `element_ptr{,_mut}`

Since `containing_record{,_mut}` is the only use of these functions, they can
be inlined to simplify things. As a bonus, this also makes the implementation
nicely parallel with the `Nt{,SingleListHead}::entry` function which goes the
other way.

DropDemBits commented:
Addressed both outstanding suggestions (took me a bit to figure out how to word the commit messages 😅).

ColinFinck reviewed (approved):
Looks good to me! Thank you again for taking care of this 👍

ColinFinck commented:
Now released as 0.2.1 and 0.3.0
